File: macos/AirRelayKit/Sources/AirRelayKit/Protocol/Messages.swift
import Foundation

public struct Hello: Codable, Sendable {
    public var protocolVersion: Int
    public var deviceName: String
    public var platform: String
    public var appVersion: String

    public init(deviceName: String, appVersion: String) {
        self.protocolVersion = 1
        self.deviceName = deviceName
        self.platform = "macos"
        self.appVersion = appVersion
    }
}

public struct NotificationPayload: Codable, Sendable, Identifiable {
    public var key: String
    public var packageName: String
    public var appName: String
    public var title: String
    public var text: String
    public var postedAt: Int64
    public var canReply: Bool
    public var actions: [String]?
    public var iconPng: String?

    public var id: String { key }
}

public struct NotificationDismiss: Codable, Sendable {
    public var key: String

    public init(key: String) {
        self.key = key
    }
}

public struct NotificationReply: Codable, Sendable {
    public var key: String
    public var text: String

    public init(key: String, text: String) {
        self.key = key
        self.text = text
    }
}

public struct NotificationAction: Codable, Sendable {
    public var key: String
    public var action: String

    public init(key: String, action: String) {
        self.key = key
        self.action = action
    }
}

public struct ClipboardText: Codable, Sendable {
    public var text: String
    public var ts: Int64

    public init(text: String, ts: Int64) {
        self.text = text
        self.ts = ts
    }
}

public struct CallState: Codable, Sendable {
    public var callId: String
    public var state: String
    public var displayName: String?
    public var number: String?
}

public struct CallAction: Codable, Sendable {
    public var callId: String
    public var action: String

    public init(callId: String, action: String) {
        self.callId = callId
        self.action = action
    }
}

public struct CameraStart: Codable, Sendable {
    public var width: Int
    public var height: Int
    public var fps: Int
    public var lens: String

    public init(width: Int = 1280, height: Int = 720, fps: Int = 30, lens: String = "back") {
        self.width = width
        self.height = height
        self.fps = fps
        self.lens = lens
    }
}

public struct DeviceStatus: Codable, Sendable {
    public var battery: Int
    public var charging: Bool
    public var wifiSsid: String?
}

public struct FileMeta: Codable, Sendable {
    public var transferId: String
    public var name: String
    public var size: Int64
    public var mime: String

    public init(transferId: String, name: String, size: Int64, mime: String) {
        self.transferId = transferId
        self.name = name
        self.size = size
        self.mime = mime
    }
}

File: android/app/src/main/java/in/aboobacker/airrelay/protocol/Messages.kt
package `in`.aboobacker.airrelay.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

val ProtocolJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
}

@Serializable
data class Hello(
    val protocolVersion: Int = 1,
    val deviceName: String,
    val platform: String = "android",
    val appVersion: String,
)

@Serializable
data class NotificationPayload(
    val key: String,
    val packageName: String,
    val appName: String,
    val title: String,
    val text: String,
    val postedAt: Long,
    val canReply: Boolean,
    val actions: List<String> = emptyList(),
    val iconPng: String? = null,
)

@Serializable
data class NotificationDismiss(val key: String)

@Serializable
data class NotificationReply(val key: String, val text: String)

@Serializable
data class NotificationAction(val key: String, val action: String)

@Serializable
data class ClipboardText(val text: String, val ts: Long)

@Serializable
data class OpenUrl(val url: String)

@Serializable
data class CallState(
    val callId: String,
    val state: String,
    val displayName: String? = null,
    val number: String? = null,
    val photoPng: String? = null,
)

@Serializable
data class CallAction(val callId: String, val action: String)

@Serializable
data class FileMeta(
    val transferId: String,
    val name: String,
    val size: Long,
    val mime: String,
)

@Serializable
data class CameraStart(
    val width: Int = 1280,
    val height: Int = 720,
    val fps: Int = 30,
    val lens: String = "back",
)

@Serializable
data class MediaState(
    val packageName: String? = null,
    val appName: String? = null,
    val title: String? = null,
    val artist: String? = null,
    val playing: Boolean = false,
    val artPng: String? = null,
)

@Serializable
data class MediaAction(val action: String)

@Serializable
data class DeviceStatus(
    val battery: Int,
    val charging: Boolean,
    val wifiSsid: String? = null,
    val networkType: String? = null,
    val signalLevel: Int? = null,
)
